File: packages/swift/Sources/EcollectSDK/Modules/SessionModule.swift
import Foundation

/// Manages ecollect session tokens: creation, in-memory caching, and auto-refresh.
public final class SessionModule {
    private let config: EcollectConfig
    private let http: HttpClient

    // MARK: - In-memory cache

    private var cachedToken: String?
    private var tokenExpiresAt: Date?
    /// Refresh threshold: request a new token when fewer than this many seconds remain.
    private let refreshThresholdSecs: Int = 300 // 5 minutes

    init(config: EcollectConfig, http: HttpClient) {
        self.config = config
        self.http = http
        // Pre-populate cache if SDK was initialized with a session token.
        if let token = config.sessionToken {
            self.cachedToken = token
            // We don't know the exact expiry, assume 30 min from now.
            self.tokenExpiresAt = Date().addingTimeInterval(1800)
        }
    }

    // MARK: - Public API

    /// Returns a valid session token, fetching or refreshing as needed.
    public func getSessionToken() async throws -> String {
        if let token = cachedToken, let expiry = tokenExpiresAt {
            let remaining = expiry.timeIntervalSinceNow
            if remaining > Double(refreshThresholdSecs) {
                return token
            }
        }
        return try await refreshToken()
    }

    /// Forces a new token to be fetched from ecollect.
    @discardableResult
    public func refreshToken() async throws -> String {
        guard let apiKey = config.apiKey else {
            throw EcollectError.invalidConfig(
                "ApiKey is required to obtain a SessionToken. " +
                "Use EcollectClient.fromSessionToken(_:) for mobile initialization."
            )
        }

        let body = GetSessionTokenRequest(
            EntityCode: config.entityCode,
            ApiKey: apiKey
        )
        let response: GetSessionTokenResponse = try await http.post(
            endpoint: "getSessionToken",
            body: body
        )

        guard response.ReturnCode == "SUCCESS", let token = response.SessionToken else {
            throw EcollectError.from(returnCode: response.ReturnCode)
        }

        cachedToken = token
        let lifetime = response.LifetimeSecs ?? 1800
        tokenExpiresAt = Date().addingTimeInterval(TimeInterval(lifetime))
        return token
    }

    /// Clears the cached session token.
    public func clearCache() {
        cachedToken = nil
        tokenExpiresAt = nil
    }

    /// Exposes the remaining lifetime of the cached token in seconds, or nil if not cached.
    public var tokenRemainingSeconds: Int? {
        guard let expiry = tokenExpiresAt else { return nil }
        let remaining = expiry.timeIntervalSinceNow
        return remaining > 0 ? Int(remaining) : nil
    }
}

File: packages/php/src/Modules/SessionModule.php
<?php

declare(strict_types=1);

namespace Ecollect\Modules;

use Ecollect\Config;
use Ecollect\Exceptions\SessionExpiredException;
use Ecollect\Utils\ErrorMapper;
use Ecollect\Utils\HttpClient;

/** Refresh proactively when less than this many seconds remain */
const PROACTIVE_REFRESH_THRESHOLD_SECS = 300;

/**
 * Manages ecollect session tokens with in-memory caching and auto-refresh.
 */
class SessionModule
{
    /** @var array{token: string, expiresAt: int, lifetimeSecs: int}|null */
    private $cached = null;

    /** @var Config */
    private $config;

    /** @var HttpClient */
    private $http;

    public function __construct(Config $config, HttpClient $http)
    {
        $this->config = $config;
        $this->http   = $http;
    }

    /**
     * Create (or re-use) a session token. Forces fetch of a new token from ecollect.
     *
     * @return array{token: string, expiresAt: int, lifetimeSecs: int}
     */
    public function create(): array
    {
        $this->cached = null;
        return $this->fetchNew();
    }

    /**
     * Return cached session token string, refreshing if < 300s remain.
     */
    public function getActive(): string
    {
        if ($this->cached !== null) {
            $remainingSecs = $this->cached['expiresAt'] - time();
            if ($remainingSecs > PROACTIVE_REFRESH_THRESHOLD_SECS) {
                return $this->cached['token'];
            }
            $this->cached = null;
        }

        $session = $this->fetchNew();
        return $session['token'];
    }

    /**
     * Force-invalidate cached token (e.g. after FAIL_APIEXPIREDSESSION).
     */
    public function invalidate(): void
    {
        $this->cached = null;
    }

    /**
     * @return array{token: string, expiresAt: int, lifetimeSecs: int}
     */
    private function fetchNew(): array
    {
        $url  = $this->config->baseUrl() . '/getSessionToken';
        $body = [
            'EntityCode' => $this->config->etyCode,
            'ApiKey'     => $this->config->apiKey,
        ];

        $res = $this->http->post($url, $body);

        if (($res['ReturnCode'] ?? '') !== 'SUCCESS' || empty($res['SessionToken'])) {
            throw ErrorMapper::map($res['ReturnCode'] ?? 'FAIL_SYSTEM', "getSessionToken failed");
        }

        $lifetimeSecs = (int)($res['LifetimeSecs'] ?? 1800);

        $session = [
            'token'        => $res['SessionToken'],
            'lifetimeSecs' => $lifetimeSecs,
            'expiresAt'    => time() + $lifetimeSecs,
        ];

        $this->cached = $session;
        return $session;
    }
}
